module/state_synchronization/indexer/persister_test.go

```diff
@@ -0,0 +1,309 @@
+package indexer
+
+import (
+	"testing"
+
+	"github.com/onflow/flow-go/storage"
+
+	"github.com/rs/zerolog"
+	"github.com/stretchr/testify/assert"
+	"github.com/stretchr/testify/mock"
+	"github.com/stretchr/testify/suite"
+
+	"github.com/onflow/flow-go/model/flow"
+	storagemock "github.com/onflow/flow-go/storage/mock"
+	"github.com/onflow/flow-go/storage/store/inmemory/unsynchronized"
+	"github.com/onflow/flow-go/utils/unittest"
+)
+
+type PersisterSuite struct {
+	suite.Suite
+	persister              *BlockPersister
+	inMemoryRegisters      *unsynchronized.Registers
+	inMemoryEvents         *unsynchronized.Events
+	inMemoryCollections    *unsynchronized.Collections
+	inMemoryTransactions   *unsynchronized.Transactions
+	inMemoryResults        *unsynchronized.LightTransactionResults
+	inMemoryTxResultErrMsg *unsynchronized.TransactionResultErrorMessages
+	registers              *storagemock.RegisterIndex
+	events                 *storagemock.Events
+	collections            *storagemock.Collections
+	transactions           *storagemock.Transactions
+	results                *storagemock.LightTransactionResults
+	txResultErrMsg         *storagemock.TransactionResultErrorMessages
+	database               *storagemock.DB
+	executionResult        *flow.ExecutionResult
+	header                 *flow.Header
+}
+
+func TestPersisterSuite(t *testing.T) {
+	t.Parallel()
+	suite.Run(t, new(PersisterSuite))
+}
+
+func (p *PersisterSuite) SetupTest() {
+	t := p.T()
+
+	block := unittest.BlockFixture()
+	p.header = block.Header
+	p.executionResult = unittest.ExecutionResultFixture(unittest.WithBlock(&block))
+
+	p.inMemoryRegisters = unsynchronized.NewRegisters(p.header.Height)
+	p.inMemoryEvents = unsynchronized.NewEvents()
+	p.inMemoryCollections = unsynchronized.NewCollections()
+	p.inMemoryTransactions = unsynchronized.NewTransactions()
+	p.inMemoryResults = unsynchronized.NewLightTransactionResults()
+	p.inMemoryTxResultErrMsg = unsynchronized.NewTransactionResultErrorMessages()
+
+	p.registers = storagemock.NewRegisterIndex(t)
+	p.events = storagemock.NewEvents(t)
+	p.collections = storagemock.NewCollections(t)
+	p.transactions = storagemock.NewTransactions(t)
+	p.results = storagemock.NewLightTransactionResults(t)
```

Comment:
I would suggest to test with real storage module, so that it could help check the integration with the real storage.

FYI, these are [some examples](https://github.com/onflow/flow-go/blob/master/storage/store/transaction_results_test.go#L22) of how setup tests with real storage modules. And instead of asserting the BatchStore method is called with certain arguments, you could check if the storing data can be retrieved with `Get` or `ByID` method.

Reply:
Yeah, I usually create tests that assert its behavior if possible, instead of checking its internal logic, because internal logic might get refactored, its behavior is less likely to change. 

One of the behavior for the happy path is that after Persist, the data can be retrieved from the storage.